add-on/src/lib/ipfs-proxy/index.js

```diff
@@ -0,0 +1,153 @@
+    if (changeInfo.status !== 'complete') return
+
+    try {
+      browser.tabs.executeScript(tab.id, {
+        file: '/dist/contentScripts/ipfs-proxy/content.js',
+        runAt: 'document_start',
+        allFrames: true
+      })
+    } catch (err) {
+      console.error('Failed to execute IPFS proxy content script', err)
+    }
+  }
+
+  browser.tabs.onUpdated.addListener(onTabUpdated)
+
+  // Port connection events are emitted when a content script attempts to
+  // communicate with us. Each new URL visited by the user will open a port.
+  // When a port is opened, we create a new IPFS proxy server to listen to the
+  // messages.
+  const onPortConnect = (port) => {
+    if (port.name !== 'ipfs-proxy') return
+
+    const { origin } = new URL(port.sender.url)
+
+    const proxy = createProxyServer(getIpfs, {
+      addListener: (_, handler) => port.onMessage.addListener(handler),
+      removeListener: (_, handler) => port.onMessage.removeListener(handler),
+      postMessage: (data) => port.postMessage(data),
+      getMessageData: (d) => d,
+      pre: ACL_FUNCTIONS.reduce((obj, permission) => {
+        obj[permission] = createAclPreCall(accessControl, origin, permission)
+        return obj
+      }, {})
+    })
+
+    const close = () => {
+      closeProxyServer(proxy)
+      port.onDisconnect.removeListener(onDisconnect)
+      connections = connections.filter(c => c.close !== close)
+    }
+
+    const onDisconnect = () => close()
+
+    // If the port disconnects, clean up the resources for this connection
+    port.onDisconnect.addListener(onDisconnect)
+
+    // Keep track of the open connections so that if destroy is called we can
+    // clean up properly.
+    connections.push({ close })
+  }
+
+  browser.runtime.onConnect.addListener(onPortConnect)
+
+  const handle = {
+    destroy () {
+      connections.forEach(c => c.destroy)
+      browser.tabs.onUpdated.removeListener(onTabUpdated)
+      browser.runtime.onConnect.removeListener(onPortConnect)
+    }
+  }
+
+  return handle
+}
+
+module.exports = createIpfsProxy
+
+// Creates a "pre" function that is called prior to calling a real function
+// on the IPFS instance. It will throw if access is denied, and ask the user if
+// no access decision has been made yet.
+function createAclPreCall (accessControl, origin, permission) {
+  return async (...args) => {
+    let access = await accessControl.getAccess(origin, permission)
+
+    if (!access) {
+      const { allow, blanket, remember } = await requestAccess(origin, permission)
+      access = await accessControl.setAccess(origin, blanket ? '*' : permission, allow, remember)
+    }
+
+    if (!access.allow) throw new Error(`User denied access to ${permission}`)
+
+    return args
+  }
+}
+
+async function requestAccess (origin, permission) {
+  const msg = `Allow ${origin} to access ipfs.${permission}?`
+
+  // TODO: add checkbox to allow all for this origin
+  let allow
+
+  try {
+    allow = window.confirm(msg)
```

Comment:
This is just FYI, but just like you said this won't work on Firefox. Modals are not allowed from background page. 
Confirmed with `ipfs.add` – it results in no prompt, access is automatically denied:

> ![2018-01-24-231332_401x57_scrot](https://user-images.githubusercontent.com/157609/35360906-949b3276-015f-11e8-9b57-376ee0fd6c43.png)
>  ![2018-01-24-231415_281x122_scrot](https://user-images.githubusercontent.com/157609/35360909-98aeba22-015f-11e8-944c-f85db0963f65.png)

Browser Console confirms prompt was not allowed:

>   ![2018-01-24-231606_835x137_scrot](https://user-images.githubusercontent.com/157609/35360919-a4e5760a-015f-11e8-8591-e2b83e4a9cdf.png)

There are related tickets at bugzilla **tl;dr** this is by design and they suggest to display prompt via content script:

> We don't intend to support alert() in background pages.
**If you want to alert data based on a context menu item, you can do so from a content script.**
> – https://bugzilla.mozilla.org/show_bug.cgi?id=1203394#c41

> I'm telling you that it won't be accessible from background pages. Alerts are modal dialogs, and they're meant to be tied to a specific, user-visible window. If you want to trigger an alert from a context menu operation, it needs to be triggered from the appropriate content window.
> – https://bugzilla.mozilla.org/show_bug.cgi?id=1203394#c43

Reply:
Yep, as I also mentioned in the description, the workaround for Firefox right now will be to go to the "manage permissions" page and change the "deny" to "allow".

I will send a separate PR for a custom confirm dialog with options to allow/deny all.

Reply:
I've noticed and found how Greasemonkey displays confirmation dialog (under Firefox):

https://github.com/greasemonkey/greasemonkey/blob/9ce5dabc74496fa6a631d991302ad4fd80730f1d/src/bg/user-script-detect.js#L76-L99

What is also interesting is Android detection via `chrome.runtime.getPlatformInfo` :)